In [1]:
from shapely.geometry import Point
import pandas as pd
import pickle
import matplotlib.pyplot as plt
from datetime import date, timedelta, datetime
import geopandas as gpd

loading may data - pickle and csv


In [2]:
%%time
july_loc = pd.read_pickle("../data/july.pkl")
july_trip = pd.read_csv("../data/july_trip.csv")

Wall time: 5.01 s


In [3]:
july_trip['pubTimeStamp'] = pd.to_datetime(july_trip['pubTimeStamp'])


This section makes a dictionary of dataframes. The dataframes can be "chunked up" into different time segments. The main things you need to adjust: the start and end variables, and the delta variable, which designates how "big" a "time chunk" should be.

In [4]:
#make a list of datetimes with datetime function 
start = datetime(2019, 7, 6, 0, 0)
end = datetime(2019, 7, 7, 0, 0)
delta = timedelta(days=1)
timelist = []
timenamelist = []
while start <= end:
    timelist.append(start)
    timenamelist.append(start.strftime("%Y-%m-%d-%H-%M"))
    start += delta
# print(timelist)
# print(timenamelist)

In [5]:
#goal of function: using list of time frames, in this case, days, make a dataframe containing
#info about the location for each scooter, for each day


#make dictionary of dataframes using base dataframe and a user-defined segments_of_time_list
#from cell above
def make_day_df(datetime_list, timenames_list, dataframe, colname):
    resultslist = {} #turn into dictionary
    x = 0
    y = 1
    z = 0
    while x in range(0,len(datetime_list)-1):
        starttime = datetime_list[x]
        endtime = datetime_list[y]
        mask = (dataframe[colname] > starttime) & (dataframe[colname] <= endtime)
        framename = dataframe.loc[mask]
        x += 1
        y += 1
        print(x)
        resultslist[timenames_list[z]] = framename
        z += 1
    return resultslist
        
the_output = make_day_df(timelist, timenamelist, july_trip,"pubTimeStamp")


1


In [6]:
the_output.keys()

dict_keys(['2019-07-06-00-00'])

I'm interested in May 26th specifically, the Saturday of Memorial Day weekend, since it had the most scooter rides.

In [7]:
the_output['2019-07-06-00-00']
# the_output['2019-05-01-00-00']['sumdid'].nunique()

pubTimeStamp companyName tripRecordNum         sumdID  \
23989 2019-07-06 00:00:08.400        Lyft          LFT1  Powered120606   
23990 2019-07-06 00:00:16.080        Lyft          LFT2  Powered852889   
23991 2019-07-06 00:00:37.000      Gotcha          GOT1    Powered2367   
23992 2019-07-06 00:00:38.007        Lyft          LFT3  Powered131810   
23993 2019-07-06 00:00:49.097        Lyft          LFT4  Powered122977   
...                       ...         ...           ...            ...   
32060 2019-07-06 23:59:26.363        Bird       BRD7616   PoweredYVI65   
32061 2019-07-06 23:59:26.363        Bird       BRD7618   PoweredXH9DD   
32062 2019-07-06 23:59:26.363        Bird       BRD7614   PoweredU7KRD   
32063 2019-07-06 23:59:26.363        Bird       BRD7615   Powered2MQXF   
32064 2019-07-06 23:59:26.363        Bird       BRD7617   Powered3JRM3   

       tripDuration  tripDistance            startDate        startTime  \
23989     21.625550   9025.590840  2019-07-05 00:00:00  23:38:30.720000   
23990      9.482850   6335.302040  2019-07-05 00:00:00  23:50:47.010000   
23991      3.000000    514.675213  2019-07-06 00:00:00         00:00:40   
23992     10.481733   6335.302040  2019-07-05 00:00:00  23:50:08.920000   
23993      9.127350    738.189000  2019-07-05 00:00:00  23:51:41.363333   
...             ...           ...                  ...              ...   
32060     26.000000  18044.620000  2019-07-06 00:00:00  23:34:56.266666   
32061     58.000000  39698.164000  2019-07-06 00:00:00  23:01:57.523333   
32062     16.000000   5249.344000  2019-07-06 00:00:00  23:43:41.390000   
32063     25.000000   5577.428000  2019-07-06 00:00:00  23:35:57.280000   
32064     32.000000  23622.048000  2019-07-06 00:00:00  23:27:55.426666   

                   endDate          endTime  startLatitude  startLongitude  \
23989  2019-07-06 00:00:00  00:00:08.253333       36.15094      -86.784230   
23990  2019-07-06 00:00:00  00:00:15.980000       36.16437      -86.779270   
23991  2019-07-06 00:00:00         00:03:40       36.16168      -86.779315   
23992  2019-07-06 00:00:00  00:00:37.826666       36.16415      -86.779670   
23993  2019-07-06 00:00:00  00:00:49.006666       36.16114      -86.777190   
...                    ...              ...            ...             ...   
32060  2019-07-07 00:00:00  00:00:57.516666       36.19010      -86.772400   
32061  2019-07-07 00:00:00  00:00:01.986666       36.16040      -86.775000   
32062  2019-07-06 00:00:00  23:59:41.600000       36.16240      -86.774600   
32063  2019-07-07 00:00:00  00:00:44.703333       36.14090      -86.794800   
32064  2019-07-06 00:00:00  23:59:44.823333       36.17420      -86.711900   

       endLatitude  endLongitude  \
23989    36.169310    -86.798320   
23990    36.176760    -86.787360   
23991    36.161789    -86.779638   
23992    36.176730    -86.787390   
23993    36.160190    -86.778760   
...            ...           ...   
32060    36.190000    -86.772200   
32061    36.160100    -86.779000   
32062    36.163900    -86.775400   
32063    36.150400    -86.796100   
32064    36.165600    -86.730400   

                                               tripRoute  \
23989  [[36.15094,-86.78423],[36.15099,-86.78424],[36...   
23990  [[36.16437,-86.77927],[36.16441,-86.77932],[36...   
23991  [["36.1618","-86.77950000000001"],["36.1620600...   
23992  [[36.16415,-86.77967],[36.16425,-86.77979],[36...   
23993  [[36.16114,-86.77719],[36.16117,-86.77723],[36...   
...                                                  ...   
32060  [[36.190122,-86.772388],[36.190131,-86.772496]...   
32061  [[36.160411,-86.774981],[36.160539,-86.775109]...   
32062    [[36.162425,-86.774499],[36.163931,-86.775322]]   
32063  [[36.14092,-86.794773],[36.140964,-86.794734],...   
32064  [[36.165591,-86.729019],[36.166145,-86.72921],...   

                        create_dt  
23989  2019-07-07 09:51:09.230000  
23990  2019-07-07 09:51:09.263000  
23991  2019-07-07 07:10:25.7

In [8]:
the_output['2019-07-06-00-00']['sumdID'].value_counts()
# the_output['2019-05-01-00-00']['sumdid'].nunique()

PoweredUNKNOWN          26
PoweredYTND2AVKAXKSI    15
PoweredFK44MOS4X2Y76    14
PoweredHEKRYNEB4AMZY    13
Powered892995           13
                        ..
PoweredMIXNW             1
PoweredFWI2W             1
PoweredP2DQ2             1
PoweredFYJPZ             1
Powered3JRM3             1
Name: sumdID, Length: 2656, dtype: int64

2630 unique scooters this day (subtracted unknowns)

In [9]:
the_output['2019-07-06-00-00']['sumdID'].value_counts().loc[lambda x : x<3]

PoweredWRTN1     2
PoweredL51XC     2
Powered502227    2
Powered953918    2
Powered742       2
                ..
PoweredMIXNW     1
PoweredFWI2W     1
PoweredP2DQ2     1
PoweredFYJPZ     1
Powered3JRM3     1
Name: sumdID, Length: 1341, dtype: int64

1341 had fewer than 3 rides. 1315 had 3 or more, or 50%.

In [10]:
the_output['2019-07-06-00-00']['sumdID'].value_counts().loc[lambda x : x>5]

PoweredUNKNOWN          26
PoweredYTND2AVKAXKSI    15
PoweredFK44MOS4X2Y76    14
PoweredHEKRYNEB4AMZY    13
Powered892995           13
                        ..
Powered991401            6
Powered874268            6
PoweredMJYTSRKL7J3UW     6
Powered869301            6
Powered552055            6
Name: sumdID, Length: 329, dtype: int64

303 had 6 or more rides. 12%

In [11]:
#make a list of datetimes with datetime function 
start = datetime(2019, 7, 25, 0, 0)
end = datetime(2019, 7, 26, 0, 0)
delta = timedelta(days=1)
timelist = []
timenamelist = []
while start <= end:
    timelist.append(start)
    timenamelist.append(start.strftime("%Y-%m-%d-%H-%M"))
    start += delta
# print(timelist)
# print(timenamelist)

In [12]:
#goal of function: using list of time frames, in this case, days, make a dataframe containing
#info about the location for each scooter, for each day


#make dictionary of dataframes using base dataframe and a user-defined segments_of_time_list
#from cell above
def make_day_df(datetime_list, timenames_list, dataframe, colname):
    resultslist = {} #turn into dictionary
    x = 0
    y = 1
    z = 0
    while x in range(0,len(datetime_list)-1):
        starttime = datetime_list[x]
        endtime = datetime_list[y]
        mask = (dataframe[colname] > starttime) & (dataframe[colname] <= endtime)
        framename = dataframe.loc[mask]
        x += 1
        y += 1
        print(x)
        resultslist[timenames_list[z]] = framename
        z += 1
    return resultslist
        
the_output = make_day_df(timelist, timenamelist, july_trip,"pubTimeStamp")


1


In [13]:
the_output.keys()

dict_keys(['2019-07-25-00-00'])

In [14]:
the_output['2019-07-25-00-00']
# the_output['2019-05-01-00-00']['sumdid'].nunique()

pubTimeStamp companyName tripRecordNum         sumdID  \
103750 2019-07-25 00:01:05.680        Lyft          LFT1  Powered583424   
103751 2019-07-25 00:02:19.287        Lyft          LFT2  Powered459446   
103752 2019-07-25 00:02:41.847        Lyft          LFT3  Powered493989   
103753 2019-07-25 00:02:49.590        Lyft          LFT4  Powered982090   
103754 2019-07-25 00:02:55.850        Lyft          LFT5  Powered089301   
...                        ...         ...           ...            ...   
107141 2019-07-25 23:59:56.603        Bird       BRD1886   PoweredU8GCD   
107142 2019-07-25 23:59:56.603        Bird       BRD1885   Powered6SN6J   
107143 2019-07-25 23:59:56.603        Bird       BRD1887   PoweredHSFX2   
107144 2019-07-25 23:59:56.603        Bird       BRD1888   PoweredYUL2B   
107145 2019-07-25 23:59:56.603        Bird       BRD1884   Powered9Y3SG   

        tripDuration  tripDistance            startDate        startTime  \
103750      2.453850    1476.37800  2019-07-24 00:00:00  23:58:38.110000   
103751      7.697450    4908.13664  2019-07-24 00:00:00  23:54:37.330000   
103752     24.080933    4698.16288  2019-07-24 00:00:00  23:38:36.760000   
103753      4.613867    1686.35176  2019-07-24 00:00:00  23:58:12.666666   
103754      9.109717    4908.13664  2019-07-24 00:00:00  23:53:49.180000   
...              ...           ...                  ...              ...   
107141      8.000000       0.00000  2019-07-25 00:00:00  23:52:06.260000   
107142      8.000000    4265.09200  2019-07-25 00:00:00  23:52:10.946666   
107143     15.000000    4265.09200  2019-07-25 00:00:00  23:46:00.420000   
107144     16.000000       0.00000  2019-07-25 00:00:00  23:44:54.143333   
107145      1.000000       0.00000  2019-07-26 00:00:00  00:00:17.996666   

                    endDate          endTime  startLatitude  startLongitude  \
103750  2019-07-25 00:00:00  00:01:05.343333       36.17713       -86.75580   
103751  2019-07-25 00:00:00  00:02:19.176666       36.14310       -86.79181   
103752  2019-07-25 00:00:00  00:02:41.616666       36.15241       -86.79140   
103753  2019-07-25 00:00:00  00:02:49.496666       36.16829       -86.78243   
103754  2019-07-25 00:00:00  00:02:55.763333       36.14314       -86.79187   
...                     ...              ...            ...             ...   
107141  2019-07-26 00:00:00  00:00:05.400000       36.14090       -86.80870   
107142  2019-07-26 00:00:00  00:00:26.690000       36.15980       -86.78050   
107143  2019-07-26 00:00:00  00:01:05.250000       36.16340       -86.78310   
107144  2019-07-26 00:00:00  00:00:56.383333       36.16410       -86.78270   
107145  2019-07-26 00:00:00  00:01:28.020000       36.15480       -86.77710   

        endLatitude  endLongitude  \
103750     36.17784     -86.75208   
103751     36.14935     -86.80201   
103752     36.15203     -86.80332   
103753     36.16560     -86.77844   
103754     36.14928     -86.80206   
...             ...           ...   
107141     36.14110     -86.80830   
107142     36.17090     -86.78520   
107143     36.15430     -86.78150   
107144     36.15300     -86.77950   
107145     36.15470     -86.77710   

                                                tripRoute  \
103750  [[36.17713,-86.7558],[36.17724,-86.75576],[36....   
103751  [[36.1431,-86.79181],[36.14314,-86.79185],[36....   
103752  [[36.15241,-86.7914],[36.15241,-86.79149],[36....   
103753  [[36.16829,-86.78243],[36.16824,-86.78229],[36...   
103754  [[36.14314,-86.79187],[36.14317,-86.79182],[36...   
...                                                   ...   
107141  [[36.140894,-86.808752],[36.141008,-86.808669]...   
107142  [[36.15986,-86.78042],[36.159773,-86.780308],[...   
107143  [[36.163659,-86.783349],[36.163656,-86.783351]...   
107144  [[36.164168,-86.782609],[36.163977,-86.782444]...   
107145  [[36.154804,-86.777045],[36.154785,-86.777061]...   

                         create_dt  
103750  2019-07-26 11:00:21.553000  
103

In [15]:
the_output['2019-07-25-00-00']['sumdID'].value_counts()
# the_output['2019-05-01-00-00']['sumdid'].nunique()

Powered013009           11
PoweredEDBDXLOZJAP6Z    10
PoweredFR3FK5ZFAH6GV    10
PoweredJ4W5YLZEMBDX2     9
Powered2QRBDKCVO7WMG     9
                        ..
Powered9VGFW             1
Powered2UVN8             1
PoweredUG5TH             1
PoweredUF71S             1
PoweredYUL2B             1
Name: sumdID, Length: 1435, dtype: int64

In [16]:
the_output['2019-07-25-00-00']['sumdID'].value_counts().loc[lambda x : x<3]

PoweredDLTTJ                                    2
Powered-05fedb39-a00f-a5e0-2219-8c3c7dd1b80e    2
Powered-d0ce22ca-e941-4fd6-55f2-d4eb9d187d22    2
PoweredUWKW3                                    2
Powered3987919                                  2
                                               ..
Powered9VGFW                                    1
Powered2UVN8                                    1
PoweredUG5TH                                    1
PoweredUF71S                                    1
PoweredYUL2B                                    1
Name: sumdID, Length: 939, dtype: int64

35% had 3 or more

In [17]:
the_output['2019-07-25-00-00']['sumdID'].value_counts().loc[lambda x : x>5]

Powered013009           11
PoweredEDBDXLOZJAP6Z    10
PoweredFR3FK5ZFAH6GV    10
PoweredJ4W5YLZEMBDX2     9
Powered2QRBDKCVO7WMG     9
                        ..
PoweredJTXV42MMK5N2K     6
PoweredOLVUVOGILUDP5     6
Powered3YI3UZR3YQTLX     6
PoweredT7ZTXTMBFPADA     6
PoweredC373KUYMR64SA     6
Name: sumdID, Length: 80, dtype: int64

6% performed well with at least 6 rides that day.

In [18]:
#make a list of datetimes with datetime function 
start = datetime(2019, 7, 10, 0, 0)
end = datetime(2019, 7, 11, 0, 0)
delta = timedelta(days=1)
timelist = []
timenamelist = []
while start <= end:
    timelist.append(start)
    timenamelist.append(start.strftime("%Y-%m-%d-%H-%M"))
    start += delta
# print(timelist)
# print(timenamelist)

In [19]:
#goal of function: using list of time frames, in this case, days, make a dataframe containing
#info about the location for each scooter, for each day


#make dictionary of dataframes using base dataframe and a user-defined segments_of_time_list
#from cell above
def make_day_df(datetime_list, timenames_list, dataframe, colname):
    resultslist = {} #turn into dictionary
    x = 0
    y = 1
    z = 0
    while x in range(0,len(datetime_list)-1):
        starttime = datetime_list[x]
        endtime = datetime_list[y]
        mask = (dataframe[colname] > starttime) & (dataframe[colname] <= endtime)
        framename = dataframe.loc[mask]
        x += 1
        y += 1
        print(x)
        resultslist[timenames_list[z]] = framename
        z += 1
    return resultslist
        
the_output = make_day_df(timelist, timenamelist, july_trip,"pubTimeStamp")


1


In [20]:
the_output.keys()

dict_keys(['2019-07-10-00-00'])

In [21]:
the_output['2019-07-10-00-00']
# the_output['2019-05-01-00-00']['sumdid'].nunique()

pubTimeStamp companyName tripRecordNum          sumdID  \
42305 2019-07-10 00:00:40.880        Lyft          LFT1   Powered658324   
42306 2019-07-10 00:01:08.693        Lyft          LFT2   Powered258676   
42307 2019-07-10 00:01:11.083        Lyft          LFT3   Powered302645   
42308 2019-07-10 00:01:26.883        SPIN        SPI184  Powered7979460   
42309 2019-07-10 00:01:26.883        SPIN        SPI183  Powered1590780   
...                       ...         ...           ...             ...   
44652 2019-07-10 23:58:48.093        Bird       BRD1450    PoweredRVNYX   
44653 2019-07-10 23:58:48.093        Bird       BRD1448    PoweredWE6DN   
44654 2019-07-10 23:58:48.093        Bird       BRD1452    Powered64W2E   
44655 2019-07-10 23:58:48.093        Bird       BRD1454    Powered3ET2X   
44656 2019-07-10 23:59:53.000      Gotcha         GOT19      Powered375   

       tripDuration  tripDistance            startDate        startTime  \
42305      6.910900   4012.467320  2019-07-09 00:00:00  23:53:46.110000   
42306      4.670317   4750.656320  2019-07-09 00:00:00  23:56:28.353333   
42307      1.666817   1000.656200  2019-07-09 00:00:00  23:59:30.893333   
42308      8.000000   2273.622120  2019-07-09 00:00:00         23:51:12   
42309     10.000000   2175.196920  2019-07-09 00:00:00         23:49:48   
...             ...           ...                  ...              ...   
44652     10.000000   6889.764000  2019-07-10 00:00:00  23:49:03.830000   
44653      4.000000   1312.336000  2019-07-10 00:00:00  23:55:52.370000   
44654     15.000000      0.000000  2019-07-10 00:00:00  23:45:24.333333   
44655     15.000000   6889.764000  2019-07-10 00:00:00  23:45:01.166666   
44656     12.000000   1121.922608  2019-07-10 00:00:00         23:59:57   

                   endDate          endTime  startLatitude  startLongitude  \
42305  2019-07-10 00:00:00  00:00:40.766666      36.156650      -86.783660   
42306  2019-07-10 00:00:00  00:01:08.573333      36.143130      -86.805480   
42307  2019-07-10 00:00:00  00:01:10.900000      36.164750      -86.779640   
42308  2019-07-09 00:00:00         23:59:05      36.153659      -86.783981   
42309  2019-07-09 00:00:00         23:58:57      36.153725      -86.784130   
...                    ...              ...            ...             ...   
44652  2019-07-10 00:00:00  23:59:23.860000      36.140800      -86.794800   
44653  2019-07-10 00:00:00  23:59:34.530000      36.152500      -86.792100   
44654  2019-07-11 00:00:00  00:00:06.026666      36.162400      -86.779600   
44655  2019-07-11 00:00:00  00:00:31.153333      36.165700      -86.780200   
44656  2019-07-11 00:00:00         00:11:57      36.149962      -86.814472   

       endLatitude  endLongitude  \
42305    36.162910    -86.776830   
42306    36.147010    -86.793920   
42307    36.167110    -86.781280   
42308    36.158838    -86.788279   
42309    36.158392    -86.788718   
...            ...           ...   
44652    36.151100    -86.784600   
44653    36.153900    -86.787600   
44654    36.151400    -86.796000   
44655    36.152000    -86.779300   
44656    36.150186    -86.810750   

                                               tripRoute  \
42305  [[36.15665,-86.78366],[36.15667,-86.78374],[36...   
42306  [[36.14313,-86.80548],[36.14316,-86.8054],[36....   
42307  [[36.16475,-86.77964],[36.16489,-86.77973],[36...   
42308  [[36.1588382,-86.7882792],[36.15835,-86.78905]...   
42309  [[36.158392277738614,-86.788717876715452],[36....   
...                                                  ...   
44652  [[36.140869,-86.794716],[36.140884,-86.794702]...   
44653  [[36.152524,-86.792008],[36.152471,-86.791896]...   
44654  [[36.162699,-86.779694],[36.162572,-86.779328]...   
44655  [[36.165712,-86.780108],[36.165399,-86.780206]...   
44656  [["36.14996","-86.81447"],["36.14996","-86.814...   

                        create_dt  
42305  2019-07-11 10:45:18.270000  
42306  2019-07-11 10:45:18.300000  
42307  2019-07-1

In [22]:
the_output['2019-07-10-00-00']['sumdID'].value_counts()
# the_output['2019-05-01-00-00']['sumdid'].nunique()

PoweredSDNMKADCNDCLN    11
Powered859498            9
PoweredLOEKT4TNO234U     9
PoweredCA5X75CSAD7HX     9
Powered7034091           9
                        ..
Powered799T4             1
Powered2AQ14             1
Powered021910            1
Powered8ZWYE             1
Powered375               1
Name: sumdID, Length: 1267, dtype: int64

1267 scooters total

In [23]:
the_output['2019-07-10-00-00']['sumdID'].value_counts().loc[lambda x : x<3]

Powered042825     2
Powered643348     2
Powered689035     2
Powered230380     2
Powered5955447    2
                 ..
Powered799T4      1
Powered2AQ14      1
Powered021910     1
Powered8ZWYE      1
Powered375        1
Name: sumdID, Length: 998, dtype: int64

269 had 3 or more. That's 21%

In [24]:
the_output['2019-07-10-00-00']['sumdID'].value_counts().loc[lambda x : x>5]

PoweredSDNMKADCNDCLN                            11
Powered859498                                    9
PoweredLOEKT4TNO234U                             9
PoweredCA5X75CSAD7HX                             9
Powered7034091                                   9
PoweredUHDEIYYDKBLZB                             8
Powered070905                                    8
PoweredB266JN56GCRG5                             8
PoweredVSEHXDME3WR37                             8
Powered940487                                    8
PoweredUADBT5UFY3FHV                             7
Powered3HJPPUNVYGL45                             7
Powered6802307                                   7
PoweredLCGAGM7XUUZCI                             7
PoweredZ72S7KBKVZWU7                             7
PoweredC6PDDC6OOF65G                             6
Powered302645                                    6
PoweredTVB3YTZZ53TBE                             6
Powered-fab97df0-bf80-4f5a-6f27-84bfb0e12033     6
Powered5BMTTDARQ2M6P           

26 scooters had 6 or more rides, or 2%